src/main/java/ecdar/backend/BackendDriver.java

```diff
@@ -122,23 +124,23 @@ private void tryStartNewEngineConnection(Engine engine) {
                 // If the process is not alive, it failed while starting up, try again
             } while (!p.isAlive());
         } else {
-            // Filter open connections to this backend and map their used ports to an int stream
-            var activeEnginePorts = startedEngineConnections.stream()
-                    .mapToInt((bi) -> Integer.parseInt(bi.getStub().getChannel().authority().split(":", 2)[1]));
+            // Filter open connections to this engine and map their used ports to an int stream
+            // and use supplier to reuse the stream for each check
+            Supplier<Stream<Integer>> activeEnginePortsStream = () -> startedEngineConnections.stream()
+                    .filter(ec -> ec.getEngine().equals(engine))
+                    .mapToInt(EngineConnection::getPort).boxed();
 
             int currentPort = engine.getPortStart();
-            do {
-                // Find port not already connected to
-                int tempPortNumber = currentPort;
-                if (activeEnginePorts.noneMatch((i) -> i == tempPortNumber)) {
-                    portNumber = tempPortNumber;
-                } else {
-                    currentPort++;
+            for (int port = engine.getPortStart(); port <= engine.getPortEnd(); port++) {
+                int tempPort = port;
+                if (activeEnginePortsStream.get().noneMatch((i) -> i == tempPort)) {
```

Comment:
I think this could be simplified by removing ``tempPort`` and using the loop variable ``port`` in the ``noneMatch`` predicate parameter. E.g. ``if (activeEnginePortsStream.get().noneMatch((i) -> i == port))``.

Reply:
This is (unfortunately) not possible in Java, as stream methods require variables to be practically final, which is not the case with `port`, even though it does not change in the scope of the loop iteration or the stream method. There is probably some ingenious explanation as to why not having the variable be final could destroy the entire system